File: blueprints/messaging/app/web/middleware/security.go
package middleware

import (
	"github.com/go-mizu/mizu"
)

// SecurityConfig holds configuration for security headers.
type SecurityConfig struct {
	// Dev indicates development mode (relaxed CSP)
	Dev bool

	// FrameOptions controls X-Frame-Options header
	// Default: "DENY"
	FrameOptions string

	// ContentTypeNoSniff enables X-Content-Type-Options: nosniff
	// Default: true
	ContentTypeNoSniff bool

	// XSSProtection enables legacy X-XSS-Protection header
	// Default: true
	XSSProtection bool

	// ReferrerPolicy sets the Referrer-Policy header
	// Default: "strict-origin-when-cross-origin"
	ReferrerPolicy string

	// CSP sets the Content-Security-Policy header
	// Default: strict policy for production
	CSP string

	// HSTS sets Strict-Transport-Security header
	// Only set in production with TLS
	HSTSMaxAge int
}

// DefaultSecurityConfig returns the default security configuration.
func DefaultSecurityConfig(dev bool) SecurityConfig {
	cfg := SecurityConfig{
		Dev:                dev,
		FrameOptions:       "DENY",
		ContentTypeNoSniff: true,
		XSSProtection:      true,
		ReferrerPolicy:     "strict-origin-when-cross-origin",
		HSTSMaxAge:         31536000, // 1 year
	}

	if dev {
		// Relaxed CSP for development
		cfg.CSP = "default-src * 'unsafe-inline' 'unsafe-eval' data: blob:; " +
			"script-src * 'unsafe-inline' 'unsafe-eval'; " +
			"style-src * 'unsafe-inline'"
	} else {
		// Strict CSP for production
		cfg.CSP = "default-src 'self'; " +
			"script-src 'self' 'unsafe-inline' 'unsafe-eval' https://cdn.tailwindcss.com https://cdnjs.cloudflare.com; " +
			"style-src 'self' 'unsafe-inline' https://fonts.googleapis.com; " +
			"img-src 'self' data: blob: https:; " +
			"font-src 'self' data: https://fonts.gstatic.com; " +
			"connect-src 'self' wss: https:; " +
			"media-src 'self' blob:; " +
			"object-src 'none'; " +
			"frame-ancestors 'none'; " +
			"base-uri 'self'; " +
			"form-action 'self'"
	}

	return cfg
}

// SecurityHeaders returns a middleware that sets security-related HTTP headers.
func SecurityHeaders(cfg SecurityConfig) mizu.Middleware {
	return func(next mizu.Handler) mizu.Handler {
		return func(c *mizu.Ctx) error {
			h := c.Writer().Header()

			// Prevent clickjacking
			if cfg.FrameOptions != "" {
				h.Set("X-Frame-Options", cfg.FrameOptions)
			}

			// Prevent MIME type sniffing
			if cfg.ContentTypeNoSniff {
				h.Set("X-Content-Type-Options", "nosniff")
			}

			// XSS protection for legacy browsers
			if cfg.XSSProtection {
				h.Set("X-XSS-Protection", "1; mode=block")
			}

			// Referrer policy
			if cfg.ReferrerPolicy != "" {
				h.Set("Referrer-Policy", cfg.ReferrerPolicy)
			}

			// Content Security Policy
			if cfg.CSP != "" {
				h.Set("Content-Security-Policy", cfg.CSP)
			}

			// Permissions Policy (formerly Feature-Policy)
			// Allow microphone for voice messages
			h.Set("Permissions-Policy",
				"geolocation=(), "+
					"microphone=(self), "+
					"camera=(), "+
					"payment=(), "+
					"usb=(), "+
					"magnetometer=(), "+
					"gyroscope=(), "+
					"accelerometer=()")

			// HSTS - only for HTTPS connections
			if !cfg.Dev && c.Request().TLS != nil && cfg.HSTSMaxAge > 0 {
				h.Set("Strict-Transport-Security",
					"max-age=31536000; includeSubDomains; preload")
			}

			// Prevent caching of sensitive data
			// Apply to API routes
			if isAPIRoute(c.Request().URL.Path) {
				h.Set("Cache-Control", "no-store, no-cache, must-revalidate, private")
				h.Set("Pragma", "no-cache")
				h.Set("Expires", "0")
			}

			return next(c)
		}
	}
}

// isAPIRoute checks if the path is an API route.
func isAPIRoute(path string) bool {
	return len(path) >= 4 && path[:4] == "/api"
}

// CORS returns a middleware that sets CORS headers.
func CORS(allowedOrigins []string, dev bool) mizu.Middleware {
	return func(next mizu.Handler) mizu.Handler {
		return func(c *mizu.Ctx) error {
			origin := c.Request().Header.Get("Origin")
			if origin == "" {
				return next(c)
			}

			allowed := false
			for _, o := range allowedOrigins {
				if origin == o {
					allowed = true
					break
				}
			}

			// In dev mode, allow localhost
			if dev && !allowed {
				devOrigins := []string{
					"http://localhost",
					"http://localhost:8080",
					"http://127.0.0.1",
					"http://127.0.0.1:8080",
				}
				for _, o := range devOrigins {
					if origin == o {
						allowed = true
						break
					}
				}
			}

			if allowed {
				h := c.Writer().Header()
				h.Set("Access-Control-Allow-Origin", origin)
				h.Set("Access-Control-Allow-Credentials", "true")
				h.Set("Access-Control-Allow-Methods", "GET, POST, PUT, PATCH, DELETE, OPTIONS")
				h.Set("Access-Control-Allow-Headers", "Content-Type, Authorization, X-CSRF-Token")
				h.Set("Access-Control-Max-Age", "86400")
				h.Set("Vary", "Origin")
			}

			// Handle preflight requests
			if c.Request().Method == "OPTIONS" {
				c.Writer().WriteHeader(204)
				return nil
			}

			return next(c)
		}
	}
}

File: blueprints/search/app/worker/src/middleware/security.ts
/**
 * Security headers middleware for Cloudflare Workers.
 * Adds standard security headers to all responses.
 */

import { createMiddleware } from 'hono/factory';

export interface SecurityHeadersConfig {
  /** Enable/disable Content-Security-Policy header (default: true) */
  contentSecurityPolicy?: boolean | string;
  /** Enable/disable X-Frame-Options header (default: true) */
  frameOptions?: boolean | 'DENY' | 'SAMEORIGIN';
  /** Enable/disable X-Content-Type-Options header (default: true) */
  contentTypeOptions?: boolean;
  /** Enable/disable X-XSS-Protection header (default: true) */
  xssProtection?: boolean;
  /** Enable/disable Referrer-Policy header (default: true) */
  referrerPolicy?: boolean | string;
  /** Enable/disable Strict-Transport-Security header (default: true in production) */
  hsts?: boolean | { maxAge?: number; includeSubDomains?: boolean; preload?: boolean };
  /** Enable/disable Permissions-Policy header (default: false) */
  permissionsPolicy?: boolean | string;
}

const DEFAULT_CSP = [
  "default-src 'self'",
  "script-src 'self' 'unsafe-inline'",  // Required for some inline scripts
  "style-src 'self' 'unsafe-inline'",   // Required for inline styles
  "img-src 'self' data: https:",        // Allow images from HTTPS sources
  "font-src 'self' data:",
  "connect-src 'self' https:",          // Allow API calls to HTTPS
  "frame-ancestors 'none'",             // Prevent embedding in frames
  "base-uri 'self'",
  "form-action 'self'",
].join('; ');

const DEFAULT_PERMISSIONS_POLICY = [
  'accelerometer=()',
  'camera=()',
  'geolocation=()',
  'gyroscope=()',
  'magnetometer=()',
  'microphone=()',
  'payment=()',
  'usb=()',
].join(', ');

/**
 * Create security headers middleware with the given configuration.
 *
 * @example
 * ```typescript
 * // Basic usage with defaults
 * app.use('*', securityHeaders());
 *
 * // Custom configuration
 * app.use('*', securityHeaders({
 *   contentSecurityPolicy: "default-src 'self'",
 *   frameOptions: 'SAMEORIGIN',
 * }));
 * ```
 */
export function securityHeaders(config: SecurityHeadersConfig = {}) {
  const {
    contentSecurityPolicy = true,
    frameOptions = true,
    contentTypeOptions = true,
    xssProtection = true,
    referrerPolicy = true,
    hsts = true,
    permissionsPolicy = false,
  } = config;

  return createMiddleware(async (c, next) => {
    await next();

    // X-Content-Type-Options - Prevent MIME type sniffing
    if (contentTypeOptions) {
      c.header('X-Content-Type-Options', 'nosniff');
    }

    // X-Frame-Options - Prevent clickjacking
    if (frameOptions) {
      const value = typeof frameOptions === 'string' ? frameOptions : 'DENY';
      c.header('X-Frame-Options', value);
    }

    // X-XSS-Protection - Enable browser XSS filter
    if (xssProtection) {
      c.header('X-XSS-Protection', '1; mode=block');
    }

    // Referrer-Policy - Control referrer information
    if (referrerPolicy) {
      const value = typeof referrerPolicy === 'string'
        ? referrerPolicy
        : 'strict-origin-when-cross-origin';
      c.header('Referrer-Policy', value);
    }

    // Content-Security-Policy - Prevent XSS and injection attacks
    if (contentSecurityPolicy) {
      const value = typeof contentSecurityPolicy === 'string'
        ? contentSecurityPolicy
        : DEFAULT_CSP;
      c.header('Content-Security-Policy', value);
    }

    // Strict-Transport-Security - Force HTTPS
    if (hsts) {
      let value: string;
      if (typeof hsts === 'object') {
        const maxAge = hsts.maxAge ?? 31536000; // 1 year default
        const parts = [`max-age=${maxAge}`];
        if (hsts.includeSubDomains) parts.push('includeSubDomains');
        if (hsts.preload) parts.push('preload');
        value = parts.join('; ');
      } else {
        value = 'max-age=31536000; includeSubDomains';
      }
      c.header('Strict-Transport-Security', value);
    }

    // Permissions-Policy - Restrict browser features
    if (permissionsPolicy) {
      const value = typeof permissionsPolicy === 'string'
        ? permissionsPolicy
        : DEFAULT_PERMISSIONS_POLICY;
      c.header('Permissions-Policy', value);
    }

    // Additional security headers
    c.header('X-Permitted-Cross-Domain-Policies', 'none');
    c.header('Cross-Origin-Opener-Policy', 'same-origin');
    c.header('Cross-Origin-Resource-Policy', 'same-origin');
  });
}

/**
 * Security headers preset for API-only endpoints (no CSP needed).
 */
export function apiSecurityHeaders() {
  return securityHeaders({
    contentSecurityPolicy: false,  // Not needed for JSON APIs
    frameOptions: true,
    contentTypeOptions: true,
    xssProtection: false,          // Not applicable to JSON
    referrerPolicy: true,
    hsts: true,
    permissionsPolicy: false,
  });
}

/**
 * Relaxed security headers for development.
 */
export function devSecurityHeaders() {
  return securityHeaders({
    contentSecurityPolicy: false,
    frameOptions: false,
    contentTypeOptions: true,
    xssProtection: false,
    referrerPolicy: true,
    hsts: false,
    permissionsPolicy: false,
  });
}
